metabase/metabase
# Refactor Action Forms

> **Warning**
> This Pull request is embarrassingly large, I will try to break it up to make it more reviewable

resolves #26724 

## Description
As part of our larger forms refactoring project, we have deprecated, and hope to delete all of our old forms frameworks that rely on turning object data structures into react forms. This makes sense for most of our app's forms, but in data apps, we are explicitly allowing users to define form settings and save them, so we do need to preserve that ability outside of our old forms frameworks.  This completely removes the dependency on old form components, and replicates some of its functionality in a way that makes more sense for data apps.

- First, this updates certain form field settings types to reflect the reality of the form fields supported (e.g. there's no special email or password components for now)
- Second, we move the category input form component out of the directories that will be deleted and into the core components folder with the other non-deprecated form components
- Third, we had an existing `ActionForm` component in the action viz - it has been renamed ActionFormViz for clarity.
- Fourth, we create a new ActionForm Component that handles all rendering of forms for actions - both their display in dashboards, and their creation in the FormCreator (with drag + drop functionality and settings, etc). This new abstraction layer lets us move certain complexities into it, like transforming saved form visualization settings into form component props.
- Finally, this new ActionForm component is nice and pure, so we can add some really tight unit tests to it.

## Screenshots

One goal of this refactor is to enable making action forms executable in the FormCreator component, so this brings the view/edit versions of action forms much closer to enable that (previously, when creating a form, it wasn't even in a <form> component). Some small visual changes to how to edit form settings and select/radio options were necessary to make those features work while actually rendering the real form components within a working <form>.

display | edit
--- | ---
![Screen Shot 2022-12-09 at 4 19 42 PM](https://user-images.githubusercontent.com/30528226/206811196-ce3aa8a2-771f-4e00-99bf-8a0ef54cea6e.png) | ![Screen Shot 2022-12-09 at 4 20 04 PM](https://user-images.githubusercontent.com/30528226/206811194-d24d7ce1-f3d4-4df7-b7f4-dbce93fa6000.png)

## Testing Steps
- you should be able to create and edit custom action forms in the action creator
- you should be able to add custom action forms to pages (both inline and in modals)
- you should be able to execute custom action forms (both inline and in modals)
- you should be able to add inline and modal versions of implicit (insert/update/delete) action forms
- you should be able execute inline and modal versions of implicit action forms
- implicit edit forms should load existing values into the form
